docs/proposals/l4-ingress.md

```diff
@@ -0,0 +1,201 @@
+
+* [keepalived-vip](https://github.com/kubernetes/contrib/tree/master/keepalived-vip)
+* NGINX Ingress Controller
+  [nginx-ingress](https://github.com/kubernetes/contrib/tree/master/ingress/controllers/nginx)
+
+However these implementations rely on a ConfigMap being populated with a
+mapping of destination endpoint to service.  A single common interface for
+describing all Ingress would is highly desirable.
+
+
+## Design
+
+### TCP and UDP support
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
+        port: 53
+      backend:
+        serviceName: testdns
+        servicePort: 53
+    - udp:
+        port: 53
+      backend:
+        serviceName: testdns
+        servicePort: 53
+```
+
+While HTTP rules have host specifiers, in TCP and UDP rules they would not be
+allowed as hostname handling is specific to L5 or higher layer protocol being
+used over TCP or UDP.  TCP and UDP as protocol specifiers should be agnostic to
+higher layer protocols.
+
+### TCP with TLS
+
+Given the following secret
+
+```yaml
+apiVersion: v1
+kind: Secret
+data:
+    tls.crt: base64 encoded cert
+    tls.key: base64 encoded key
+metadata:
+    name: testsecret
+type: Opaque
+```
+
+TCP with TLS would be almost identical to the existing TLS Ingress
+specification with the addition of the tcp attribute for the Ingress rule
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
+        port: 3306
+      host: www.foo.com
+      tls:
+        - secretName: testsecret
+      backend:
+        serviceName: testmysql
+        servicePort: 3306
+```
+
+TLS as a protocol does support hostnames via SNI, thus allowing a host
+specifier seems to make sense and opens the possibility of supporting multiple
+TCP services on a single IP.
+
+### Ingress type specifier
+
+The current Ingress API assumes a protocol based on the presence or absence
+of a key (tls, http).  This leaves open the possibility of a spec that defines
+all protocols.
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
+        port: 3306
+      udp:
+        port: 3306
+      host: www.foo.com
+      tls:
+        - secretName: testsecret
+      backend:
+        serviceName: testmysql
+        servicePort: 3306
+      http:
+        paths:
+        - path: /bar
+          backend:
+              serviceName: bar
+              servicePort: 80
+```
+
+A specification like the one above is arguably too complex and should be
+written as separate rules for each protocol being supported.  However, the
+loose syntax of the specification allows construction of a complex
+specification.  Using a 'type' or 'protocol' specifier would remove this
+ambiguity in the specification.
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
```

Comment:
this again feels like multiple distinct ingresses pointing at the same "claim", instead of jamming everything into one